arroyo/dead_letter_queue/dead_letter_queue.py

```diff
@@ -0,0 +1,54 @@
+import logging
+from typing import Optional
+from arroyo.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessage,
+)
+from arroyo.processing.strategies.abstract import ProcessingStrategy as ProcessingStep
+from arroyo.types import Message, TPayload
+
+logger = logging.getLogger(__name__)
+
+
+class DeadLetterQueue(ProcessingStep[TPayload]):
+    """
+    DLQ Processing Step.
+
+    Attempts to submit a given message to the next processing step,
+    handling an `InvalidMessage` according to the given Policy.
+    """
+
+    def __init__(
+        self,
+        next_step: ProcessingStep[TPayload],
+        policy: DeadLetterQueuePolicy[TPayload],
+    ) -> None:
+        self.__next_step = next_step
+        self.__policy = policy
+        self.__closed = False
+
+    def poll(self) -> None:
+        self.__next_step.poll()
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+        try:
+            self.__next_step.submit(message)
+        except InvalidMessage as e:
+            logger.warning(
+                f"Invalid Message caught by Dead Letter Queue: {message}",
+                exc_info=True,
+            )
+            self.__policy.handle_invalid_message(message, e)
```

Comment:
When we want to start producing the invalid message to a dead letter topic, it could happen here. If the handler call doesn't raise an exception, produce the message to the dead letter topic.

Is there a case where we do not raise an exception and also do not want the message to be forwarded? If so, the handler could return a boolean telling the main class to forward the message or not.

Reply:
It's more intuitive to me to have the DLQ be responsible for receiving and then forwarding the message, and the policy possibly interrupts this process by raising an exception.

This way the DLQ could be initialized with a topic to produce to, it feels off to have the policy initialized with a topic to produce a message to. The "produce to dead letter topic" logic would also have to be repeated across policies (unless a base Policy class implements this)